### DISCUSSION TITLE
Confused with "runs-on" and "container" options

### DISCUSSION THREAD

**Discussion by bek0s** [community/community#25534] [Actions]
<p>I have just started experimenting with the CI/CD functionality of GitHub Actions, and I fail to understand the “runs-on” and “container” options.</p>
<ul>
<li>
<p>Why both “runs-on” and “container” are required when I want to run a container, and what exactly does this mean? Let’s say for example that I have “runs-on: ubuntu-latest” and “container: centos-latest”. Does this mean that my requested centos-latest image will run on an ubuntu-latest virtual machine? Why do I even have to specify the host operating system (ubuntu-latest) if I plan to use a container and do everything I want to do there?</p>
</li>
<li>
<p>If I try “runs-on: macos-latest” and “container: centos-latest”, at the job setup I get the following error: “##[error]File not found: ‘docker’”, which make sense because MacOS can’t run Docker. But why is there an option to have a CI configuration file like this?</p>
</li>
<li>
<p>What if someone wants to create a build matrix with different containers? To my understanding this is not possible at the moments because the build matrix somehow works only with the “runs-on” option.</p>
</li>
<li>
<p>Wouldn’t it be better if there was no “container” option, and instead one would have to specify the docker image on the “runs-on” option?</p>
</li>
</ul>
<p>Please excuse me if I am getting something totally the wrong way… I started experimenting with CI/CD only a couple of days ago.</p>
<p>Thanks in advance.</p>

**Comment by ethomson** [comment, marked as answer]
<p>&gt; Why both “runs-on” and “container” are required when I want to run a container, and what exactly does this mean? Let’s say for example that I have “runs-on: ubuntu-latest” and “container: centos-latest”. Does this mean that my requested centos-latest image will run on an ubuntu-latest virtual machine? Why do I even have to specify the host operating system (ubuntu-latest) if I plan to use a container and do everything I want to do there?</p>
<p>Yes, you’re exactly right - that’s what that means.</p>
<p>You need to specify that because the underlying virtual machine dictates the type of containers that you can run.  You need an underlying Windows virtual machine (ie, <code>runs-on: windows-latest</code>) to use Windows based containers.  And you need an underlying Linux virtual machine (<code>runs-on: ubuntu-latest</code>) to use Linux based containers.</p>
<p>Since we don’t know what type of container you’re trying to run until we download it, we can’t let you skip specifying a virtual machine type.  (We need to provision a virtual machine just to do that download.)</p>
<p>&gt; What if someone wants to create a build matrix with different containers? To my understanding this is not possible at the moments because the build matrix somehow works only with the “runs-on” option.</p>
<p>That’s possible - there’s nothing special about the runs-on key in a matrix.  A matrix is really only variable substitution.  To use a matrix of different containers, you can:</p>
<pre><code>jobs:
  build:
    name: Build on ${{matrix.container }}
    runs-on: ubuntu-latest
    container: ${{ matrix.container }}
    strategy:
      matrix:
        container: ['ubuntu:16.04', 'ubuntu:18.04']
    steps:
    - run: cat /etc/debian_version
</code></pre>

**Comment by bek0s** [comment]
<p>Excellent explanation! Thank you so much my friend!</p>

**Reply by ihsansfd** [reply]
you're not his friend tho

**Comment by fullstackfool** [comment]
<p>Searching for exactly this explination - many thanks for explaining so cleanly</p>

**Comment by MeenalJy** [comment]
> Why both “runs-on” and “container” are required when I want to run a container, and what exactly does this mean? Let’s say for example that I have “runs-on: ubuntu-latest” and “container: centos-latest”. Does this mean that my requested centos-latest image will run on an ubuntu-latest virtual machine? Why do I even have to specify the host operating system (ubuntu-latest) if I plan to use a container and do everything I want to do there?

Before jumping to the WHY question, let's first dig into the WHAT and the essence of these two. With [**runs-on**], we can specify the host environment of the workflow. We can specify the job environment with [**container**]. Back to the original question, we are enhancing task isolation and efficiently managing different environments with a simple configuration file.

![Ubuntu-latest](https://github.com/community/community/assets/107105216/75720c7c-1e1a-4ddd-8a76-8f9edee582d5)


> If I try “runs-on: macos-latest” and “container: centos-latest”, at the job setup I get the following error: “##[error]File not found: ‘docker’”, which make sense because MacOS can’t run Docker. But why is there an option to have a CI configuration file like this?

You are getting this error because the kernels of MacOS and CentOS are different. The docker container uses the kernel of the host machine and should belong to the same Operating System. 


> Wouldn’t it be better if there was no “container” option, and instead one would have to specify the docker image on the “runs-on” option?

If you will use only "runs-on" then you will have only 1 Operating System i.e. Host OS, but if you use it along with "container" you will get different environment ( Host OS, Container OS) to work with and you can simply create individual jobs in your workflow file, each specifying a different container.
